astropy/modeling/tests/test_quantities_evaluation.py

```diff
@@ -0,0 +1,538 @@
+
+        @staticmethod
+        def evaluate(x, a):
+            return a * x
+
+    m = TestModelB(a=1 / u.s)
+    with pytest.raises(UnitsError) as exc:
+        m(3 * u.m)
+    assert exc.value.args[0] == ("Units of output 'y', m / s (speed), could not be "
+                                 "converted to required output units of m (length)")
+
+
+def test_output_units_functional():
+    """
+    Test setting output_units to a custom function that returns a unit.
+    """
+
+    class TestModelC(Fittable1DModel):
+        a = Parameter()
+
+        # In this case the output's units are some compound
+        # involving both the input and parameter's units
+        output_units = lambda a, x: a.unit * x.unit
+
+        @staticmethod
+        def evaluate(x, a):
+            return a * x
+
+    m = TestModelC(a=1 / u.s)
+    assert m(0).unit == (1 / u.s)
+    assert m(2).unit == (1 / u.s)
+    assert m(0 * u.dimensionless_unscaled).unit == (1 / u.s)
+    assert m(2 * u.dimensionless_unscaled).unit == (1 / u.s)
+    assert m(2 * u.m).unit == (u.m / u.s)
+
+
+@pytest.mark.xfail
+def test_output_units_functional_inconsistent():
+    """
+    Test setting output_units to a custom function that returns units
+    inconsistent with the output of the evaluate method.
+    """
+
+    class TestModelC(Fittable1DModel):
+        a = Parameter()
+
+        # In this case the output's units are some compound
+        # involving both the input and parameter's units
+        output_units = lambda a, x: a.unit * x.unit
+
+        @staticmethod
+        def evaluate(x, a):
+            return a * x * u.s
+
+    m = TestModelC(a=1 / u.s)
+    with pytest.raises(UnitsError) as exc:
+        m(3 * u.m)
+    assert exc.value.args[0] == ("Units of output 'y', m (length), could not be "
+                                 "converted to required output units of m / s "
+                                 "(speed)")
+
+
+def test_output_units_fixed_unit():
+    """
+    Test setting output_units to a fixed unit.
+    """
+
+    class TestModelD(Fittable1DModel):
+
+        # This is a no-op model that just always forces the output to be in
+        # meters (if the input is a length)
+        output_units = u.m
+
+        @staticmethod
+        def evaluate(x):
+            return 2 * x
+
+    m = TestModelD()
+    assert m(0).unit is u.m
+    assert m(1 * u.m) == 2 * u.m
+    assert m(1 * u.km) == 2000 * u.m
+
+
+def test_output_units_fixed_unit_inconsistent():
+    """
+    Test setting output_units to a fixed unit that is inconsistent with
+    evaluate method output.
+    """
+
+    class TestModelD(Fittable1DModel):
+
+        # This is a no-op model that just always forces the output to be in
+        # meters (if the input is a length)
+        output_units = u.m
+
+        @staticmethod
+        def evaluate(x):
+            return 2 * x
+
+    m = TestModelD()
+    assert m(0).unit is u.m
+    assert m(1 * u.m) == 2 * u.m
+    assert m(1 * u.km) == 2000 * u.m
+
```

Comment:
Is this case consistent with the docstring, which suggests that it is testing a case where the supplied input will produce something that is not consistent with the specified output unit? But these cases are consistent.